### ISSUE TITLE
Reevaluate using Babel over TypeScript

### ISSUE THREAD

**Issue by JulianKniephoff** [elan-ev/tobira#39]
We currently use Babel to translate our TypeScript to JavaScript. This is cool because of multiple reasons (see below), but it is also inconsistent with some of the choices we made. We use the TypeScript ESLint plugin for linting, for example, which uses the TypeScript compiler internally, and we also use the TypeScript compiler directly ourselves to do type checking. This could potentially lead to weird behavior where these tools understand different features.

We **should** be fine if we just use TypeScript, because presumably all TS tooling should understand that. And if we agree on this, the current tooling setup might actually be a good choice because of the combination of the advantages of both approaches (see below), although we might want to think about enforcing the "only standard TypeScript" rule somehow, for example using Babel configuration and/or linter rules.

If we choose to unify the tooling (use TS compiler over babel OR use Babel based linting and ??? for type checking), here is a rough overview over that trade off:

Babel > TypeScript:

* Potentially more cool modern ES (or not-quite-yet-ES, aka proposed ES) features!
* We could even use crazier stuff that is not just for ES backwards compatibility, like `babel-plugin-macros`
* More control over the output
  * TS let's you specify a target ES version, Babel let's you turn on/off individual features, we have `browserslist` integration, ...

TypeScript > Babel

* Better integration of type checking and the build process
  * In the "Babel only case" I don't know how to get type checking
  * In our current setup, type checking is an independent step; if the code is syntactically correct but doesn't type check, we (can) still get a build
    * This might actually be good sometimes for prototyping, but overall I think it is kind of defeating the purpose of a type checker
* The TypeScript based linting can incorporate types in its rules

# TL;DR

We should traverse the following decision tree:

1. Unify our transpiler tooling to avoid weird compatibility traps?
   1. Base everything on Babel?
      1. How to do type checking?!
   2. Base everything on TypeScript?
2. Stick with our current setup, and potentially find different ways to do so?
   1. Find ways to restrict our code to things that both, Babel and TypeScript understand?
   2. Actually we could also consider switching **just** the linter from TS to Babel, or even use **both**, but I guess this decision can still be deferred to when we find good reasons.

**Comment by JulianKniephoff** [comment]
I personally would favor 1.ii over 1.i because of the type checking, but I like 2.i even more. As for 2.ii, I don't care, yet.

**Comment by LukasKalbertodt** [comment]
- I unfortunately also favor 2.i. I really dislike having this mix of many different build tools, but AFAIK babel is just better for output generation compared to TS. And AFAIK, type checking (which we obviously want somehow) is only possible with TS right now. 
- Regarding linters: I don't mind more linters.
- Making sure that we only use code understood by TS and Babel is a good idea. But shouldn't this be checked by CI already as that typechecks AND runs babel?
- I agree that creating a build without type checking is... meh. But since typechecking and building are two separate steps, typechecking will make the build notably slower, which is a shame. But I wouldn't mind modifying the `build:*` scripts to include typechecking. Most of the time, we will use other tools (file watchers, ...) to do the building for us anyway.

**Comment by JulianKniephoff** [comment]
> * Making sure that we only use code understood by TS and Babel is a good idea. But shouldn't this be checked by CI already as that typechecks AND runs babel?

That's a fair point, I didn't think about that. :see_no_evil:

**Comment by LukasKalbertodt** [comment]
@JulianKniephoff  What's the state here? Are you still considering to switch somehow? Otherwise, we can maybe close this?

**Comment by JulianKniephoff** [comment]
No. :)
